[CMake] [Do Not Merge] Allow building non-stdlib Swift libraries

harlanhaskins opened the pull request:
The only thing that previously informed the module install directory was
the presence of the IS_STDLIB flag when calling add_swift_library. This
adds a MODULE_DIR directory through the add_swift_library chain such
that a non-stdlib module can be built.

@gottesmm: I can't figure out how to actually compute a `MODULE_DIR` for `swiftSwiftSyntax` because the only way to compute the `LIBRARY_SUBDIR` that `IS_STDLIB` infers requires an SDK and architecture, which isn't known when declaring the library. Should this be _another_ flag like `BUILD_ALONGSIDE_STDLIB` or something?

harlanhaskins committed:
Allow building non-stdlib Swift libraries

The only thing that previously informed the module install directory was
the presence of the IS_STDLIB flag when calling add_swift_library. This
adds a MODULE_DIR directory through the add_swift_library chain such
that a non-stdlib module can be built.

jrose-apple commented on cmake/modules/AddSwift.cmake:
```diff
@@ -580,6 +581,9 @@ endfunction()
 #   Install library dylib and swift module files to lib/swift.
 #
+# MODULE_DIR
+#   Directory to install .swiftmodule and .o file in.
```
.o files don't get installed.

harlanhaskins replied:
I was mirroring the comment for `MODULE_DIR` here, but you're right `install` is the wrong word. https://github.com/apple/swift/blob/17f619793829b1ccc351b956c6b8f2ad618dc16f/cmake/modules/SwiftSource.cmake#L149-L150

jrose-apple reviewed (commented):
Shouldn't TARGET_LIBRARY be sufficient here?

harlanhaskins commented:
@jrose-apple Should we do the inference we use for `IS_STDLIB` but for `TARGET_LIBRARY` instead? That makes sense.

harlanhaskins committed:
Use same comment for MODULE_DIR as in _compile_swift_file

harlanhaskins commented on cmake/modules/SwiftSource.cmake:
```diff
@@ -78,6 +78,17 @@ function(handle_swift_sources
+    precondition(NOT SWIFTSOURCES_IS_STDLIB OR NOT SWIFTSOURCES_MODULE_DIR,
+      "Cannot provide MODULE_DIR if IS_STDLIB is set")
+
+    if(SWIFTSOURCES_IS_STDLIB)
```
@jrose-apple Would it be sufficient to change `if(SWIFTSOURCES_IS_STDLIB)` to `if(SWIFTSOURCES_TARGET_LIBRARY)` and pass the `TARGET_LIBRARY` param into `handle_swift_sources`?

jrose-apple replied:
Honestly, I have no idea. What TARGET_LIBRARYs are not IS_STDLIB?

harlanhaskins replied:
Right now, SwiftSyntax.

harlanhaskins replied:
Also swiftImageInspectionShared, swiftRuntime, swiftReflection, swiftStdlibStubs, and swiftRemoteMirror

jrose-apple replied:
None of those other things have any Swift in them, do they.

harlanhaskins replied:
Right, no they don't.

harlanhaskins replied:
Okay, so I tried this for a while -- making `TARGET_LIBRARY` infer `MODULE_DIR` instead of `IS_STDLIB`. CMake, in turn, stopped building `SwiftSyntax` entirely. Then I moved it from `tools/` to `stdlib/public/`, and it built just fine.

There seems to be some more implicit behavior (maybe because `tools/` is included before `stdlib/`?) that's preventing this.

harlanhaskins commented:
@gottesmm: I'd really like to figure this out so we can un-disable the SwiftSyntax tests and actually ship the library.

gottesmm commented:
@harlanhaskins I gotta focus on some stuff now. That being said, where were things left off, you never got back to me (no worries).

harlanhaskins commented:
Okay so where we left off:

I refactored MODULE_DIR up into handle_swift_sources and add_swift_library, then didn’t have a way to compute the MODULE_DIR, so I just made TARGET_LIBRARY do the inference that MODULE_DIR used to. This still didn’t work with SwiftSyntax in the tools directory — only when I moved it to stdlib/public.

harlanhaskins commented:
Closing this in favor of #11752.
